src/librustc/ty/mod.rs

```diff
@@ -2077,31 +2079,43 @@ impl<'a, 'gcx, 'tcx> AdtDef {
            repr: ReprOptions) -> Self {
         debug!("AdtDef::new({:?}, {:?}, {:?}, {:?})", did, kind, variants, repr);
         let mut flags = AdtFlags::NO_ADT_FLAGS;
+
+        if kind == AdtKind::Enum && tcx.has_attr(did, "non_exhaustive") {
+            debug!("found non-exhaustive variant list for {:?}", did);
+            flags = flags | AdtFlags::IS_VARIANT_LIST_NON_EXHAUSTIVE;
+        }
+        flags |= match kind {
+            AdtKind::Enum => AdtFlags::IS_ENUM,
+            AdtKind::Union => AdtFlags::IS_UNION,
+            AdtKind::Struct => AdtFlags::IS_STRUCT,
+        };
+
+        if let AdtKind::Struct = kind {
+            let variant_def = &variants[VariantIdx::new(0)];
+            let def_key = tcx.def_key(variant_def.did);
+            match def_key.disambiguated_data.data {
+                DefPathData::StructCtor => flags |= AdtFlags::IS_TUPLE_STRUCT,
```

Comment:
Unit structs have constructors and support `Self` values as well, so something like `AdtFlags::HAS_CTOR` would be more correct.

Reply:
> Two commits ([be8a579](https://github.com/rust-lang/rust/commit/be8a5791d8b5696e76346b49cfa886e715bd680c), [b1a193d](https://github.com/rust-lang/rust/commit/b1a193d90b22e6d81a10a455802332ffbd413ccb)) are not directly related to this PR.

Sure, but one is pretty closely related, so we might as well include it... the cosmetic one, up to you, but I don't see why not. :-)

Reply:
Makes sense.